Screen 4 active deal

BartDeCaluwe opened the pull request:
This PR will add:

- Deals are now saved in store, so we can prevent RPC calls when navigating to a deal
- Deal screen
- DealDetail component and Storybook story
- DealAspect component and Storybook stories
- The admin status is saved in the store based on the connected wallet
- Pending deals are now shown based on admin status
- Activate deal button is show based on admin status

BartDeCaluwe committed:
feat: get marketplace from route instead of from consts

BartDeCaluwe committed:
feat: add invest button

BartDeCaluwe committed:
feat: add deal details

BartDeCaluwe committed:
feat: add pending deals

BartDeCaluwe committed:
fix: deal detail link

BartDeCaluwe committed:
feat: add centered layout

BartDeCaluwe committed:
feat: add pending deals

BartDeCaluwe committed:
feat: only show pending deals and activate deal button to admins

BartDeCaluwe committed:
fix: storybook build failing

BartDeCaluwe committed:
feat: add DealStatus component

BartDeCaluwe committed:
feat: add DealDetails component

BartDeCaluwe committed:
feat: fetch market from store

BartDeCaluwe committed:
feat: add link to buttons

BartDeCaluwe committed:
feat: use store for saving admin status

BartDeCaluwe committed:
feat: save deals in store

By saving the deals in the store we can prevent reloading all the deals on the deal details page

BartDeCaluwe committed:
test: fix failing test caused by bad import

BartDeCaluwe committed:
feat: add dealdetails story

BartDeCaluwe committed:
feat: use ui amount formatter for deal related amounts

BartDeCaluwe committed:
feat: redirect pending deal to dealdetails

BartDeCaluwe committed:
feat: deal detail layout

BartDeCaluwe committed:
feat: add principal repaid ratio indicator

BartDeCaluwe committed:
feat: add DealAspect component

BartDeCaluwe committed:
feat: reset deal data when setting market

This way we prevent deal data from the wrong market being shown

BartDeCaluwe committed:
fix: DealAspect ratio check failing when ratio is 0

RuvendeGroote commented on src/utils/format.utils.ts:
```diff
@@ -1,9 +1,25 @@
 import { Ratio } from "@credix/credix-client";
 import Big, { RoundingMode } from "big.js";
 
+export const clamp = (value: number, min: number, max: number) => {
```
we can write tests for stuff like this (or use the lodash clamp function)

RuvendeGroote commented on src/utils/format.utils.ts:
```diff
@@ -1,9 +1,25 @@
 const roundingPrecision = 2;
 const conversionFactor = new Big(10).pow(6);
 
+export const numberFormatter = Intl.NumberFormat("en", {
```
same here, utils like this can be tested

RuvendeGroote commented on src/utils/format.utils.ts:
```diff
@@ -15,8 +31,13 @@ export const formatUIAmount = (n: Big, roundingMode: RoundingMode, formatter: an
+	const numerator = new Big(r.numerator);
 	return numerator.div(denominator).mul(100);
 };
 
```
why is the fallback n-GB below?

BartDeCaluwe replied:
No particular reason, we just need **a** fallback. Would you rather have it be "en-US"?

RuvendeGroote commented on src/utils/format.utils.ts:
```diff
@@ -15,8 +31,13 @@ export const formatUIAmount = (n: Big, roundingMode: RoundingMode, formatter: an
 export const toProgramAmount = (n: Big) => n.mul(conversionFactor);
 
 export const formatRatio = (r: Ratio) => {
```
this can be implemented on the Ratio class itself

BartDeCaluwe replied:
We actually might not even need this code. We can use `Ratio.apply` here.

RuvendeGroote commented on src/state/marketSlice.ts:
```diff
@@ -1,21 +1,85 @@
 export type MarketSlice = {
 	market?: Market;
+	allDeals?: Deal[];
+	activeDeals?: Deal[];
```
the deal state arrays can be selectors on alldeals
I think we should move to separate selector files anyways

RuvendeGroote commented on src/state/marketSlice.ts:
```diff
@@ -1,21 +1,85 @@
+	getDeal: (client: CredixClient, marketPlace: string, dealAddress: string) => Promise<Deal>;
+};
+
+const getMarket = (client: CredixClient, marketPlace: string) => {
```
this function doesn't add much

RuvendeGroote commented on src/state/marketSlice.ts:
```diff
@@ -1,21 +1,85 @@
+};
+
+const getDeals = async (client: CredixClient, marketPlace: string, get: GetState<MarketSlice>) => {
+	const market = get().market ? get().market : await getMarket(client, marketPlace);
```
if we just do a maybefetch market we can then do a get afterwards and it will have the same logic but more readable and concise

RuvendeGroote commented on src/state/marketSlice.ts:
```diff
@@ -1,21 +1,85 @@
+		return;
+	}
+
+	set({ isLoadingDeals: true });
```
can we put the isLoadingDeals in the getDeals function?

RuvendeGroote commented on src/state/marketSlice.ts:
```diff
@@ -1,21 +1,85 @@
+		marketPlace,
+		get
+	);
+	set({ allDeals, activeDeals, endedDeals, pendingDeals, isLoadingDeals: false });
```
same here

RuvendeGroote commented on src/state/marketSlice.ts:
```diff
@@ -1,21 +1,85 @@
 	market: null,
+	...initialDealsState,
+	isLoadingDeals: false,
 	fetchMarket: async (client, marketPlace) => {
```
put these functions as consts somewhere so they don't clutter this object so much

RuvendeGroote replied:
also, we don't use fetchMarket it seems so let's remove it

RuvenSalamon replied:
this still applies

RuvendeGroote commented on src/state/marketSlice.ts:
```diff
@@ -25,4 +89,25 @@ export const createMarketSlice: StoreSlice<MarketSlice> = (set, get) => ({
 		const market = await getMarket(client, marketPlace);
 		set({ market });
 	},
+	fetchDeals: async (client, marketPlace) => {
```
also not used it seems

RuvendeGroote commented on src/state/marketSlice.ts:
```diff
@@ -25,4 +89,25 @@ export const createMarketSlice: StoreSlice<MarketSlice> = (set, get) => ({
+		await maybeFetchDeals(client, marketPlace, get, set);
+
+		if (!get().allDeals) {
+			return;
```
do we expect the deal to exist here? if so I would fail hard. I don't like silent failures because it hides bugs

RuvendeGroote commented on src/pages/[marketplace]/index.tsx:
```diff
@@ -6,6 +6,7 @@ import { useCredixClient } from "@credix/credix-client";
 import { useRouter } from "next/router";
 import Layout from "@components/Layout";
 import { NextPageWithLayout } from "pages/_app";
+import Link from "next/link";
```
double import

RuvendeGroote commented:
![image](https://user-images.githubusercontent.com/14349484/159240129-37c9b35c-4b4d-49a9-a392-3614d0b0ee9b.png)
this looks weird

RuvendeGroote commented on src/pages/[marketplace]/deals.tsx:
```diff
@@ -45,19 +47,21 @@ const dealsTableColumns = [
 
 const Deals: NextPage = () => {
```
add layout

RuvendeGroote commented on src/pages/[marketplace]/deals.tsx:
```diff
@@ -45,19 +47,21 @@ const dealsTableColumns = [
-	const [isLoadingDeals, setIsLoadingDeals] = useState<boolean>(true);
-	const [activeDeals, setActiveDeals] = useState<Deal[]>([]);
-	const [endedDeals, setEndedDeals] = useState<Deal[]>([]);
+	const [isAdmin, setIsAdmin] = useState<boolean>(false);
```
either use state per component or use the zustand admin state

RuvendeGroote commented:
when fetching the deals we preform the same call a lot
![image](https://user-images.githubusercontent.com/14349484/159240963-3eed0fd0-bc87-4146-9cad-88671a175389.png)

RuvendeGroote commented:
> when fetching the deals we preform the same call a lot ![image](https://user-images.githubusercontent.com/14349484/159240963-3eed0fd0-bc87-4146-9cad-88671a175389.png)

![image](https://user-images.githubusercontent.com/14349484/159241050-bce01135-c96b-4c99-bad5-f3ec0731b048.png)

BartDeCaluwe committed:
feat: add layout to deals page

BartDeCaluwe committed:
feat: add layout to deal detail page

BartDeCaluwe committed:
test: fix failing unit test

RuvendeGroote commented on src/state/marketSlice.ts:
```diff
@@ -1,21 +1,85 @@
-import { CredixClient, Market } from "@credix/credix-client";
+import { GetState, SetState } from "zustand";
```
add tests

BartDeCaluwe committed:
feat: add some rpc optimisations

Only fetch data if we're not already fetching said data

BartDeCaluwe committed:
feat: remove redundant code

the getGatekeeperNetwork function only gets the gatekeepernetwork from the market which will update itself when the object changes in the state.

BartDeCaluwe committed:
test: add clamp, numberformat and ratioformat tests

BartDeCaluwe committed:
feat: get admin state from store

BartDeCaluwe committed:
refactor: create dealslice and selectors

BartDeCaluwe committed:
fix: divide by zero error when interest to repay was 0

BartDeCaluwe committed:
feat: get deal from dealslice

BartDeCaluwe committed:
refactor: remove formatRatio in favor op Ratio.appy

BartDeCaluwe committed:
feat: add link to ended deals

BouweCeunen commented:
<img width="829" alt="image" src="https://user-images.githubusercontent.com/5576686/160351801-55a270f9-617b-4698-9e33-b6928381a972.png">
following Figma, left should be `0%` instead of just a `0` maybe? Also not sure what the `0` is on the right with "time left".

BartDeCaluwe committed:
fix: ratios with a value of 0 not getting formatted

BartDeCaluwe committed:
test: add zustand mock

BartDeCaluwe committed:
feat: add admin slice test

RuvendeGroote commented:
conflicts

BartDeCaluwe committed:
fix: number formatting

BartDeCaluwe committed:
test: add tests for dealSlice and marketSlice

BartDeCaluwe committed:
feat: dont show percentage amount for days remaining

BartDeCaluwe committed:
fix: marketstats not always loading

RuvenSalamon commented on __tests__/__mocks__/zustand.ts:
```diff
@@ -0,0 +1,29 @@
+const create = (createState) => {
+	const store = actualCreate<StoreState>(createState);
+	const initialState = store.getState();
+	storeResetFns.add(() => store.setState(initialState, true));
```
we don't have multiple stores right now as far as I know so no point in an elaborate setup to support them. KISS

RuvendeGroote commented on __tests__/__mocks__/zustand.ts:
```diff
@@ -0,0 +1,29 @@
+import { StoreState } from "@state/useStore";
+
+// a variable to hold reset functions for all stores declared in the app
+const storeResetFns = new Set();
```
we don't have multiple stores right now as far as I know so no point in an elaborate setup to support them. KISS

RuvendeGroote replied:
+ WAGNI!

RuvendeGroote commented on src/components/DealStatus.tsx:
```diff
@@ -0,0 +1,21 @@
+import { Tag } from "./Tag";
+
+interface DealStatusProps {
+	deal?: Deal;
```
why is this optional? it's only used in a place where there is always a deal

RuvendeGroote commented on src/components/DealDetails.tsx:
```diff
@@ -0,0 +1,83 @@
+
+	useEffect(() => {
+		const principalRatio = new Ratio(
+			toUIAmount(deal?.principalAmountRepaid).toNumber(),
```
deal always exists

RuvendeGroote commented on src/components/DealAspect.tsx:
```diff
@@ -0,0 +1,37 @@
+	title: string;
+	value: string;
+	ratio?: number;
+	showRatio?: boolean;
```
is there a case where we have a ratio but don't show it? or is there a reason why we would want to pass a ratio but not show it? showRatio and ratio seem tightly coupled right now so just use ratio

BartDeCaluwe replied:
The ratio is used for both the rendering of the "progress bar" and just displaying the formatted value. In the case of the amount of days left, only the progress bar is displayed, not the actual value. Hence, the distinction.

RuvendeGroote commented on src/components/DealAspect.tsx:
```diff
@@ -0,0 +1,37 @@
+interface DealAspectProps {
+	title: string;
+	value: string;
+	ratio?: number;
```
just accept a Ratio here

RuvendeGroote commented on src/components/DealDetails.tsx:
```diff
@@ -0,0 +1,83 @@
+}
+
+export const DealDetails: FunctionComponent<DealDetailsProps> = ({ deal }) => {
+	const [interestRepaidRatio, setInterestRepaidRatio] = useState<number | 0>();
```
use Ratio here

RuvendeGroote commented on src/components/DealDetails.tsx:
```diff
@@ -0,0 +1,83 @@
+		);
+		setPrincipalRepaidRatio(principalRatio.apply(new Big(1)).toNumber());
+
+		if (deal?.interestToRepay.eq(new Big(0))) {
```
if you use the ratio interestRepaid/totalInterest it will be 1 if everything is repaid and this logic can be simplified

RuvendeGroote commented on src/components/DealDetails.tsx:
```diff
@@ -0,0 +1,83 @@
+			setInterestRepaidRatio(interestRatio.apply(new Big(1)).toNumber());
+		}
+
+		if (deal?.principalAmountRepaid) {
```
we don't show it so don't need it

BartDeCaluwe committed:
feat: pass Ratios to DealAspect instead of numbers

BartDeCaluwe committed:
fix: deal is not optional

BartDeCaluwe committed:
test: simplify resetting of store after each test run

BartDeCaluwe committed:
fix: simplify marketslice and dealslice

RuvendeGroote commented on src/components/DealStatus.tsx:
```diff
@@ -0,0 +1,21 @@
+}
+
+export const DealStatus = ({ deal }: DealStatusProps) => {
+	if (!deal) {
```
if deal is not optional this check is also no longer required

RuvendeGroote commented on src/components/DealAspect.tsx:
```diff
@@ -0,0 +1,41 @@
+			{hasRatio && (
+				<div
+					className="absolute top-0 left-0 h-1 bg-neutral-60"
+					style={{ width: `${clamp(ratio.apply(new Big(1)).toNumber() * 100, 0, 100)}%` }}
```
if this is something you feel could be useful to in the Ratio class itself you can always propose it. it's basically `toPercentage()` or something

BartDeCaluwe replied:
Yes, IMO could be useful!

RuvendeGroote commented on src/state/dealSlice.ts:
```diff
@@ -0,0 +1,49 @@
+export const createDealSlice: StoreSlice<DealSlice> = (set, get) => ({
+	...initialDealsState,
+	isLoadingDeals: false,
+	fetchDeals: (market) => getDeals(market, set),
```
getDeals doesn't return `Promise<void>`
